ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java

```diff
@@ -261,45 +283,151 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+   * @param replicaCapacityInBytes the minimum amount of free space on a disk for a replica
+   * @param datacenter the {@link Datacenter} to allocate replicas in
+   * @param retryIfNotRackAware {@code true} if we should attempt a non rack-aware allocation if a rack-aware one
+   *                            is not possible.
+   * @return a list of {@code replicaCountPerDatacenter} or fewer disks that can be allocated for a new partition in
+   *         the specified datacenter
+   */
+  private List<Disk> allocateDisksForPartitionWithPotentialRetry(int replicaCountPerDatacenter,
+      long replicaCapacityInBytes, Datacenter datacenter, boolean retryIfNotRackAware) {
+    List<Disk> disks;
+    if (datacenter.isRackAware()) {
+      disks = allocateDisksForPartition(replicaCountPerDatacenter, replicaCapacityInBytes, datacenter, true);
+      if ((disks.size() < replicaCountPerDatacenter) && retryIfNotRackAware) {
+        System.err.println("Rack-aware allocation failed for a partition on datacenter:" + datacenter.getName()
+            + "; attempting to perform a non rack-aware allocation.");
+        disks = allocateDisksForPartition(replicaCountPerDatacenter, replicaCapacityInBytes, datacenter, false);
       }
+    } else if (!retryIfNotRackAware) {
+      throw new IllegalArgumentException("retryIfNotRackAware is false, but the datacenter: " + datacenter.getName()
+          + " does not have rack information");
+    } else {
+      disks = allocateDisksForPartition(replicaCountPerDatacenter, replicaCapacityInBytes, datacenter, false);
+    }
+
+    if (disks.size() < replicaCountPerDatacenter) {
+      System.err.println(
+          "Could only allocate " + disks.size() + "/" + replicaCountPerDatacenter + " replicas in datacenter: "
+              + datacenter.getName());
     }
-    return allocatedDisks;
+    return disks;
   }
 
-  // Best effort (or less) allocation of partitions. I.e., size of returned list may be less than numPartitions.
+  /**
+   * Allocate partitions for {@code numPartitions} new partitions on all datacenters.
+   *
+   * @param numPartitions How many partitions to allocate.
+   * @param replicaCountPerDatacenter The number of replicas per partition on each datacenter
+   * @param replicaCapacityInBytes How large each replica (of a partition) should be
+   * @param retryIfNotRackAware {@code true} if we should attempt a non rack-aware allocation if a rack-aware one
+   *                            is not possible.
+   * @return A list of the new {@link PartitionId}s.
+   */
   public List<PartitionId> allocatePartitions(int numPartitions, int replicaCountPerDatacenter,
-      long replicaCapacityInBytes) {
+      long replicaCapacityInBytes, boolean retryIfNotRackAware) {
     ArrayList<PartitionId> partitions = new ArrayList<PartitionId>(numPartitions);
 
     while (checkEnoughUnallocatedRawCapacity(replicaCountPerDatacenter, replicaCapacityInBytes) && numPartitions > 0) {
```

Comment:
not related to the changes in this patch, but it looks like this check does not take into account the fact that replicas of a partition cannot be on the same node (it simply checks total disk capacity of all nodes in the datacenter with total size required for all replicas).